src/passes/MemoryCopyFillLowering.cpp

```diff
@@ -0,0 +1,184 @@
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeCall(
+        "__memory_fill", {curr->dest, curr->value, curr->size}, Type::none)
+    );
+    needsMemoryFill = true;
+  }
+
+  void run(Module* module) override {
+    if (!module->features.hasBulkMemory()) {
+      return;
+    }
+    if (module->features.hasMemory64() || module->features.hasMultiMemory()) {
+      throw "Memory64 and multi-memory not supported";// TODO: best way to return an error?
+    }
+    // In order to introduce a call to a function, it must first exist, so create an empty stub.
+    Builder b(*module);
+    auto memCpyFunc = b.makeFunction("__memory_copy",
+      Signature({Type::i32, Type::i32, Type::i32}, {Type::none}), {Type::i32});
+    memCpyFunc->body = b.makeBlock();
+    module->addFunction(memCpyFunc.release());
+    auto memFillFunc = b.makeFunction("__memory_fill",
+      Signature({Type::i32, Type::i32, Type::i32}, {Type::none}), {});
+    memFillFunc->body = b.makeBlock();
+    module->addFunction(memFillFunc.release());
+
+    Super::run(module);
+
+    if (needsMemoryCopy) {
+      Index dst = 0, src = 1, size = 2, temp = 3;
+      Name memory = module->memories.front()->name;
+      Block* body = b.makeBlock();
+      // temp = memory size in bytes
+      body->list.push_back(b.makeLocalSet(temp,
+        b.makeBinary(BinaryOp::MulInt32,
+          b.makeMemorySize(memory),
+          b.makeConst(Memory::kPageSize))));
+      // if dst + size > memsize or src + size > memsize, then trap.
+      body->list.push_back(
+        b.makeIf(
+          b.makeBinary(BinaryOp::OrInt32,
+            b.makeBinary(BinaryOp::GtUInt32,
+              b.makeBinary(BinaryOp::AddInt32,
+                b.makeLocalGet(dst, Type::i32), b.makeLocalGet(size, Type::i32)),
+              b.makeLocalGet(temp, Type::i32)),
+            b.makeBinary(BinaryOp::GtUInt32,
+              b.makeBinary(BinaryOp::AddInt32,
+                b.makeLocalGet(src, Type::i32), b.makeLocalGet(size, Type::i32)),
+              b.makeLocalGet(temp, Type::i32))),
+        b.makeUnreachable())
+      );
+      // temp = dst
+      body->list.push_back(b.makeLocalSet(temp, b.makeLocalGet(dst, Type::i32)));
+      // dst = dst + size
+      body->list.push_back(
+        b.makeLocalSet(dst,
+          b.makeBinary(BinaryOp::AddInt32,
+            b.makeLocalGet(dst, Type::i32),
+            b.makeLocalGet(size, Type::i32))));
+      // src = src + size
+      body->list.push_back(
+        b.makeLocalSet(src,
+          b.makeBinary(BinaryOp::AddInt32,
+            b.makeLocalGet(src, Type::i32),
+            b.makeLocalGet(size, Type::i32))));
+      body->list.push_back(b.makeBlock("out",
+        b.makeLoop("copy", b.makeBlock({
+          // break if dst == temp
+          b.makeBreak("out", nullptr,
+            b.makeBinary(BinaryOp::EqInt32,
+              b.makeLocalGet(dst, Type::i32),
+              b.makeLocalGet(temp, Type::i32))),
+          // --dst; --src;
```

Comment:
I think we may need to pick the direction, up or down, based on the overlap? `wasm-interpreter.h` does that.